File: esj-core/src/main/java/de/bsnsoft/esj/model/Term.java
package de.bsnsoft.esj.model;

import de.bsnsoft.esj.EsjFormatException;
import de.bsnsoft.esj.SemanticType;
import de.bsnsoft.esj.TermKind;
import java.util.EnumSet;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.OptionalInt;
import java.util.Set;

/**
 * One entry of the registry: a business term or a business group, with the structural
 * facts the validation layers are defined against and the documentation members the
 * registry carries beside them (specification, section 10).
 *
 * <p>The members {@code slug}, {@code order}, {@code codeList}, {@code maxDecimals},
 * {@code maxDecimalsRule}, {@code reqIds}, {@code description} and {@code notes} are
 * documentation and code generation aids and never affect validation or the canonical
 * form.
 *
 * @param id          the identifier, for example {@code BT-131} or {@code BG-DEX-01}
 * @param kind        business term or business group
 * @param name        the name of the term in the described model
 * @param slug        a lower camel case member name for code generation
 * @param parent      the identifier of the enclosing group, or an empty optional for a
 *                    term directly at the root
 * @param path        the identifiers from the root down to and including this term
 * @param depth       the number of enclosing groups
 * @param cardinality the declared cardinality inside one instance of the parent
 * @param datatype    the semantic data type, or an empty optional for a group
 * @param maxDecimals the maximum number of fraction digits, where the edition fixes a
 *                    constant
 * @param maxDecimalsRule how the edition derives the maximum number of fraction digits
 *                    from the document, where it fixes no constant; stated in place of
 *                    {@code maxDecimals} and never beside it
 * @param codeList    the name of the code list a code is taken from, where there is one
 * @param components  the supplementary components of the semantic data type
 * @param reusesTerms the identifiers of terms of the base model that this extension group
 *                    carries one level deeper (specification, section 5.6)
 * @param order       the one-based position of the term in the table of the described model
 * @param reqIds      the requirement identifiers the described edition's term table gives
 *                    for this entry, empty where it gives none
 * @param description one sentence explaining the term, written for the registry
 * @param notes       further remarks written for the registry
 */
public record Term(String id,
                   TermKind kind,
                   String name,
                   String slug,
                   Optional<String> parent,
                   List<String> path,
                   int depth,
                   Cardinality cardinality,
                   Optional<SemanticType> datatype,
                   OptionalInt maxDecimals,
                   Optional<String> maxDecimalsRule,
                   Optional<String> codeList,
                   List<Component> components,
                   List<String> reusesTerms,
                   int order,
                   List<String> reqIds,
                   String description,
                   List<String> notes) {

    /** The components an Identifier has (EN 16931-1, 6.5.6). */
    private static final Set<Component.Role> IDENTIFIER_ROLES =
            EnumSet.of(Component.Role.SCHEME, Component.Role.SCHEME_VERSION);

    /** The components a Binary Object has, both of them mandatory (EN 16931-1, 6.5.11). */
    private static final Set<Component.Role> BINARY_OBJECT_ROLES =
            EnumSet.of(Component.Role.MIME_CODE, Component.Role.FILENAME);

    /**
     * Copies the collections, checks that every part is present, and checks the
     * components against the semantic data type: a component role belongs to exactly one
     * data type, a Binary Object carries both of its components and both are mandatory
     * whatever its component list says, and a scheme version stands only beside a scheme
     * and is mandatory only beside a mandatory one (EN 16931-1, 6.5.6 and 6.5.11;
     * specification, section 6.2). The registry is the single source of typing, so a
     * combination the model does not have is refused here rather than carried into the
     * validator, the generated schema and the typed view.
     *
     * @param id          the identifier, for example {@code BT-131} or {@code BG-DEX-01}
     * @param kind        business term or business group
     * @param name        the name of the term in the described model
     * @param slug        a lower camel case member name for code generation
     * @param parent      the identifier of the enclosing group, or an empty optional for a
     *                    term directly at the root
     * @param path        the identifiers from the root down to and including this term
     * @param depth       the number of enclosing groups
     * @param cardinality the declared cardinality inside one instance of the parent
     * @param datatype    the semantic data type, or an empty optional for a group
     * @param maxDecimals the maximum number of fraction digits, where the edition fixes a
     *                    constant
     * @param maxDecimalsRule how the edition derives the maximum number of fraction digits
     *                    from the document, where it fixes no constant; stated in place of
     *                    {@code maxDecimals} and never beside it
     * @param codeList    the name of the code list a code is taken from, where there is one
     * @param components  the supplementary components of the semantic data type
     * @param reusesTerms the identifiers of terms of the base model that this extension group
     *                    carries one level deeper (specification, section 5.6)
     * @param order       the one-based position of the term in the table of the described model
     * @param reqIds      the requirement identifiers the described edition's term table gives
     *                    for this entry, empty where it gives none
     * @param description one sentence explaining the term, written for the registry
     * @param notes       further remarks written for the registry
     * @throws NullPointerException     if a part is {@code null}
     * @throws IllegalArgumentException if the path is empty or does not end at this term
     * @throws EsjFormatException       if the components do not fit the semantic data type
     */
    public Term {
        Objects.requireNonNull(id, "id");
        Objects.requireNonNull(kind, "kind");
        Objects.requireNonNull(name, "name");
        Objects.requireNonNull(slug, "slug");
        Objects.requireNonNull(parent, "parent");
        Objects.requireNonNull(cardinality, "cardinality");
        Objects.requireNonNull(datatype, "datatype");
        Objects.requireNonNull(maxDecimals, "maxDecimals");
        Objects.requireNonNull(maxDecimalsRule, "maxDecimalsRule");
        Objects.requireNonNull(codeList, "codeList");
        Objects.requireNonNull(description, "description");
        path = List.copyOf(path);
        components = List.copyOf(components);
        reusesTerms = List.copyOf(reusesTerms);
        reqIds = List.copyOf(reqIds);
        notes = List.copyOf(notes);
        if (maxDecimals.isPresent() && maxDecimalsRule.isPresent()) {
            throw new EsjFormatException(id + " states its decimal limit as a number and"
                    + " as a rule, and two limits are two answers to one question");
        }
        if (path.isEmpty() || !path.get(path.size() - 1).equals(id)) {
            throw new IllegalArgumentException("the path of " + id + " ends at " + id);
        }
        checkComponents(id, datatype, components);
    }

    private static void checkComponents(String id,
                                        Optional<SemanticType> datatype,
                                        List<Component> components) {
        Set<Component.Role> roles = EnumSet.noneOf(Component.Role.class);
        for (Component component : components) {
            if (!roles.add(component.role())) {
                throw new EsjFormatException(id + " lists the component "
                        + component.role().jsonMember() + " twice");
            }
        }
        if (datatype.orElse(null) == SemanticType.BINARY_OBJECT) {
            checkBinaryObject(id, roles, components);
            return;
        }
        if (components.isEmpty()) {
            return;
        }
        SemanticType type = datatype.orElseThrow(() -> new EsjFormatException(
                id + " carries supplementary components but no semantic data type"));
        Set<Component.Role> allowed = switch (type) {
            case IDENTIFIER -> IDENTIFIER_ROLES;
            case BINARY_OBJECT -> BINARY_OBJECT_ROLES;
            default -> Set.of();
        };
        for (Component.Role role : roles) {
            if (!allowed.contains(role)) {
                throw new EsjFormatException("the semantic data type " + type.registryDatatype()
                        + " of " + id + " has no component " + role.jsonMember());
            }
        }
        if (roles.contains(Component.Role.SCHEME_VERSION)
                && !roles.contains(Component.Role.SCHEME)) {
            throw new EsjFormatException(
                    id + " carries a scheme version without a scheme");
        }
        if (mandatory(components, Component.Role.SCHEME_VERSION)
                && !mandatory(components, Component.Role.SCHEME)) {
            throw new EsjFormatException(id + " declares its scheme version mandatory and"
                    + " its scheme optional, and no value can satisfy both");
        }
    }

    /**
     * Checks the components of a Binary Object, whose two components are a property of
     * the semantic data type rather than of the term: both are present and both are
     * mandatory (EN 16931-1, 6.5.11; specification, section 6.2). An empty component list
     * is therefore the defect itself and not the absence of one, which is why this runs
     * before the shortcut that lets a term without components pass.
     */
    private static void checkBinaryObject(String id,
                                          Set<Component.Role> roles,
                                          List<Component> components) {
        if (!roles.equals(BINARY_OBJECT_ROLES)) {
            throw new EsjFormatException("the Binary Object " + id
                    + " carries the components mimeCode and filename");
        }
        for (Component component : components) {
            if (!component.isMandatory()) {
                throw new EsjFormatException("the component "
                        + component.role().jsonMember() + " of the Binary Object " + id
                        + " is mandatory");
            }
        }
    }

    /** Tells whether the component in that role is present and declared mandatory. */
    private static boolean mandatory(List<Component> components, Component.Role role) {
        for (Component component : components) {
            if (component.role() == role) {
                return component.isMandatory();
            }
        }
        return false;
    }

    /**
     * Tells whether this entry is a business group.
     *
     * @return {@code true} for a group
     */
    public boolean isGroup() {
        return kind == TermKind.BG;
    }

    /**
     * Tells whether this term may occur more than once inside one instance of its parent
     * and therefore carries an occurrence index in every path.
     *
     * @return {@code true} if the declared maximum is greater than 1
     */
    public boolean isRepeatable() {
        return cardinality.isRepeatable();
    }

    /**
     * Tells whether this term must be present in every instance of its parent.
     *
     * @return {@code true} if the declared minimum is at least 1
     */
    public boolean isMandatory() {
        return cardinality.isMandatory();
    }

    /**
     * Returns the supplementary component carried in a given value object member.
     *
     * @param role the member of the value object
     * @return the component, or an empty optional if the registry lists none for this term
     * @throws NullPointerException if {@code role} is {@code null}
     */
    public Optional<Component> component(Component.Role role) {
        Objects.requireNonNull(role, "role");
        for (Component component : components) {
            if (component.role() == role) {
                return Optional.of(component);
            }
        }
        return Optional.empty();
    }
}

File: bindings/csharp/En16931.SemanticJson/Model/Term.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace En16931.SemanticJson.Model;

/// <summary>Which member of an ESJ value a supplementary component feeds.</summary>
public enum ComponentRole
{
    /// <summary>The identification scheme of an identifier.</summary>
    Scheme,

    /// <summary>The version of that identification scheme.</summary>
    SchemeVersion,

    /// <summary>The media type of an attachment.</summary>
    MimeCode,

    /// <summary>The file name of an attachment.</summary>
    Filename,
}

/// <summary>
/// A supplementary component of a semantic data type, as a registry records it for one term
/// (specification, section 6.1).
/// </summary>
/// <param name="Role">which value member the component feeds</param>
/// <param name="Name">the name the model gives the component</param>
/// <param name="Min">the minimum cardinality: 1 where the registry declares it mandatory</param>
/// <param name="SchemeList">the code list the scheme comes from, where the model fixes one</param>
public sealed record Component(ComponentRole Role, string Name, int Min, string? SchemeList)
{
    /// <summary>Returns the member of an ESJ value this component feeds.</summary>
    public string JsonMember => Role switch
    {
        ComponentRole.Scheme => "scheme",
        ComponentRole.SchemeVersion => "schemeVersion",
        ComponentRole.MimeCode => "mimeCode",
        _ => "filename",
    };

    /// <summary>Tells whether the registry declares this component mandatory.</summary>
    public bool IsMandatory => Min >= 1;

    /// <summary>Returns the role a registry writes under that name.</summary>
    /// <param name="role">the <c>role</c> member of a component</param>
    /// <returns>the role</returns>
    /// <exception cref="EsjFormatException">if no role carries that name</exception>
    public static ComponentRole RoleOf(string role) => role switch
    {
        "scheme" => ComponentRole.Scheme,
        "schemeVersion" => ComponentRole.SchemeVersion,
        "mimeCode" => ComponentRole.MimeCode,
        "filename" => ComponentRole.Filename,
        _ => throw new EsjFormatException(null,
            "the registry names the component role " + role + ", which no value member carries"),
    };
}

/// <summary>How often a term or a group may occur under its parent.</summary>
/// <param name="Min">the minimum, 0 or 1</param>
/// <param name="Max">the maximum, 1 or <see cref="Unbounded"/> where the registry writes <c>n</c></param>
public sealed record Cardinality(int Min, int Max)
{
    /// <summary>The maximum a registry writes as <c>n</c>.</summary>
    public const int Unbounded = int.MaxValue;

    /// <summary>Tells whether more than one occurrence is admitted.</summary>
    public bool IsRepeatable => Max > 1;

    /// <summary>Tells whether at least one occurrence is required.</summary>
    public bool IsMandatory => Min >= 1;

    /// <inheritdoc />
    public override string ToString() =>
        Min.ToString(CultureInfo.InvariantCulture) + ".." + (Max == Unbounded ? "n" : Max.ToString(CultureInfo.InvariantCulture));
}

/// <summary>
/// One business term or business group of the semantic model, as a registry records it
/// (specification, section 10).
/// </summary>
public sealed class Term
{
    private readonly List<string> _path;
    private readonly List<Component> _components;
    private readonly List<string> _reusesTerms;

    /// <summary>Creates a term.</summary>
    /// <param name="id">the identifier of the model, used verbatim in a semantic path</param>
    /// <param name="isGroup">whether the term is a business group</param>
    /// <param name="name">the name the model gives it</param>
    /// <param name="slug">the language-neutral name stem for code generation</param>
    /// <param name="parent">the enclosing group, or <c>null</c> at the root</param>
    /// <param name="path">the identifiers from the root down to this term</param>
    /// <param name="cardinality">how often it may occur under its parent</param>
    /// <param name="datatype">the semantic data type, or <c>null</c> for a group</param>
    /// <param name="components">the supplementary components the registry lists</param>
    /// <param name="reusesTerms">terms an extension group carries one level deeper</param>
    /// <param name="codeList">the code list a code term draws from, without a version</param>
    public Term(
        string id,
        bool isGroup,
        string name,
        string slug,
        string? parent,
        IEnumerable<string> path,
        Cardinality cardinality,
        SemanticType? datatype,
        IEnumerable<Component> components,
        IEnumerable<string> reusesTerms,
        string? codeList)
    {
        Id = id ?? throw new ArgumentNullException(nameof(id));
        IsGroup = isGroup;
        Name = name;
        Slug = slug;
        Parent = parent;
        _path = new List<string>(path);
        Cardinality = cardinality;
        Datatype = datatype;
        _components = new List<Component>(components);
        _reusesTerms = new List<string>(reusesTerms);
        CodeList = codeList;
    }

    /// <summary>Returns the identifier of the model.</summary>
    public string Id { get; }

    /// <summary>Tells whether this is a business group.</summary>
    public bool IsGroup { get; }

    /// <summary>Returns the name the model gives the term.</summary>
    public string Name { get; }

    /// <summary>Returns the language-neutral name stem for code generation.</summary>
    public string Slug { get; }

    /// <summary>Returns the enclosing group, or <c>null</c> at the root of the document.</summary>
    public string? Parent { get; }

    /// <summary>Returns the identifiers from the root down to and including this term.</summary>
    public IReadOnlyList<string> Path => _path;

    /// <summary>Returns how often the term may occur under its parent.</summary>
    public Cardinality Cardinality { get; }

    /// <summary>Returns the semantic data type, or <c>null</c> for a group.</summary>
    public SemanticType? Datatype { get; }

    /// <summary>Returns the supplementary components the registry lists for the term.</summary>
    public IReadOnlyList<Component> Components => _components;

    /// <summary>Returns the terms an extension group carries one level deeper.</summary>
    public IReadOnlyList<string> ReusesTerms => _reusesTerms;

    /// <summary>Returns the code list the term draws from, or <c>null</c>.</summary>
    public string? CodeList { get; }

    /// <summary>Tells whether the term may occur more than once.</summary>
    public bool IsRepeatable => Cardinality.IsRepeatable;

    /// <summary>Tells whether at least one occurrence is required.</summary>
    public bool IsMandatory => Cardinality.IsMandatory;

    /// <summary>Returns the component of a role, or <c>null</c> where the registry lists none.</summary>
    /// <param name="role">the role</param>
    /// <returns>the component, or <c>null</c></returns>
    public Component? ComponentOf(ComponentRole role)
    {
        foreach (Component component in _components)
        {
            if (component.Role == role)
            {
                return component;
            }
        }

        return null;
    }

    /// <inheritdoc />
    public override string ToString() => Id + " " + Cardinality;
}
